### PR TITLE
Expand section on testing (gradle/guides#147)

### PR DESCRIPTION
**Pull request by ysb33r**
- Added the three testing styles
- Broken document into smaller parts and large file was becoming a pain to maintain.
- Added recommendation on listing blocks.

### CODE REVIEW

**Comment by ysb33r** [comment]
@bmuschko, @kousen You can provide feedback on this. Will keep adding on it as we go on. Planning to finally merge it after next week after feedback has been collected and more things have come out from reviewing all other guides.

Feel free to just check the branch out and push updates if you wish, it might be quicker.

File: contents/parts/commands-and-code-listings.adoc [hunk 1]
```diff
@@ -0,0 +1,43 @@
+== Commands and code listings
+
+Almost all guides include commands to run and code listings to copy, so the following guidelines are generally applicable.
```
Comment (bmuschko) [inline contents/parts/commands-and-code-listings.adoc]: I don't quite understand what this sentence is conveying. We should think about rephrasing.

File: contents/parts/make-guides-testable.adoc [hunk 2]
```diff
@@ -0,0 +1,108 @@
+== Make guides testable
+
+A testable guide is not a guide that shows readers how to add tests, but is one where the included code snippets and even the console output is validated during the building of the guide. Having guides testable, results in lower maintenance for guide authors, especially when new versions of Gradle come out. Familiarize yourself with Asciidoctor's use of link:http://asciidoctor.org/docs/user-manual/#include-directive[file include] and link:http://asciidoctor.org/docs/user-manual/#by-tagged-regions[include tags] as they play a key role for including code tested code snippets and console output.
```
Comment (bmuschko) [inline contents/parts/make-guides-testable.adoc]: This should explain the conventions introduced with https://github.com/gradle-guides/gradle-guides-plugin/pull/10. Where do we expect to find code snippets and console output with a concrete directory tree example.

File: contents/parts/make-guides-testable.adoc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+There are three specific ways to make a guide testable. As an author you can select one of more of these methods. The exact ones you select will be determined by the context of the guide you are writing.
```
Comment (bmuschko) [inline contents/parts/make-guides-testable.adoc]: Based on our discussion we should take out your plugin from the list at the moment. This portion of the style guide will have to evolve as the code changes.

File: contents/parts/make-guides-testable.adoc [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+<2> Copy required source files to the working project directory.
+<3> `workingDir` is the folder where `gradleRunner` keeps the project it is performing the step upon.
+
+NOTE: Always place a `settings.gradle` in your example project, even if it is empty.
+
+You can get more information on using `gradleRunner` in the http://ysb33r.github.io/gradleTest/[GradleRunner plugin documentation].
+
+=== Test method 2: Include a testsuite
+
+Include a test suite in the project. When the guide has many code snippets. Place the test code in the `src/test` folder and follow the Gradle conventions for writing tests.
```
Comment (bmuschko) [inline contents/parts/make-guides-testable.adoc]: Didn't this change to `src/test/guide`?

File: contents/parts/make-guides-testable.adoc [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+Here is an example of a test environment that a guide author has created as part of validating a guide
+
+.build.gradle
+[source,groovy]
+----
+test {
+    systemProperties 'HTTPSERVER_PORT' : 65000
+    systemProperties 'TEMPLATE_DIR' : file('src/publicationTest').absolutePath
+
+    inputs.dir 'src/publicationTest'  // <1>
```
Comment (bmuschko) [inline contents/parts/make-guides-testable.adoc]: We should probably bake that into the guides plugin and point to `samples` directory.

File: contents/parts/make-guides-testable.adoc [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+For your convenience apply the the `org.gradle.guides.test-jvm-code` plugin when the code under test is JVM-based. You can use the Spock Framework for the testing, but are not limited to that. Configure the tests in the usual `test` block.
+
+Here is an example of a test environment that a guide author has created as part of validating a guide
+
+.build.gradle
+[source,groovy]
+----
+test {
+    systemProperties 'HTTPSERVER_PORT' : 65000
+    systemProperties 'TEMPLATE_DIR' : file('src/publicationTest').absolutePath
```
Comment (bmuschko) [inline contents/parts/make-guides-testable.adoc]: We won't need that anymore. This is provided by the guides plugin internally.

File: contents/parts/make-guides-testable.adoc [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+----
+<1> Include `settings.gradle` and `build.gradle` from the example project.
+<2> Copy required source files to the working project directory.
+<3> `workingDir` is the folder where `gradleRunner` keeps the project it is performing the step upon.
+
+NOTE: Always place a `settings.gradle` in your example project, even if it is empty.
+
+You can get more information on using `gradleRunner` in the http://ysb33r.github.io/gradleTest/[GradleRunner plugin documentation].
+
+=== Test method 2: Include a testsuite
```
Comment (bmuschko) [inline contents/parts/make-guides-testable.adoc]: Should be "test suite" - two separate words.

File: contents/parts/make-guides-testable.adoc [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+test {
+    systemProperties 'HTTPSERVER_PORT' : 65000
+    systemProperties 'TEMPLATE_DIR' : file('src/publicationTest').absolutePath
+
+    inputs.dir 'src/publicationTest'  // <1>
+}
+----
+<1> Ensure the test ran when the sources used internally by the test changed.
+
+=== Test method 3: Include a full project for native guides
```
Comment (bmuschko) [inline contents/parts/make-guides-testable.adoc]: I don't quite understand how 3 is different from 2. Can you elaborate?

**Comment by bmuschko** [comment]
@ysb33r Is this something you are planning to fix/merge soon?

**Pull request merged**